dbms/src/Common/assert_cast.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <type_traits>
+#include <typeinfo>
+#include <typeindex>
+#include <string>
+
+#include <Common/Exception.h>
+#include <common/demangle.h>
+
+
+namespace DB
+{
+    namespace ErrorCodes
+    {
+        extern const int BAD_CAST;
+    }
+}
+
+
+/** Perform static_cast in release build.
+  * Checks type by comparing typeid and throw an exception in debug build.
+  * The exact match of the type is checked. That is, cast to the ancestor will be unsuccessful.
```

Comment:
What does it do for non-polymorphic types, just works like a static_cast?
And for polymorphic ones, it is like typeid_cast, but asserts that the result is nonzero?

Reply:
boost has a similar thing called polymorphic_cast (though I like the name 'assert_cast' more).

https://www.boost.org/doc/libs/1_71_0/doc/html/conversion/polymorphic_casts.html#polymorphic_cast

Reply:
Curiously, the list of things detected by ubsan doesn't mention invalid static_casts.

https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#available-checks

Reply:
> What does it do for non-polymorphic types, just works like a static_cast?

It will fail if types are non-identical. It means, that it is not applicable for non-polymorphic types.

Reply:
> Curiously, the list of things detected by ubsan doesn't mention invalid static_casts.

There is. Look at `ColumnVectorHelper.h`.

Reply:
> boost has a similar thing called polymorphic_cast (though I like the name 'assert_cast' more).

That looks like what we need. There is another subtle difference: we compare typeids instead of doing dynamic_cast - it is more weak.